### ISSUE TITLE
2.5.1 分布式多卡 疑似不同步模型参数

### ISSUE THREAD

**Issue by johnyanccer** [PaddlePaddle/Paddle#56807]
### bug描述 Describe the Bug

2.5.1 分布式 多卡训练 时间不同步

主要代码如下：
paddle.device.set_device('gpu')
gen = paddle.seed(int(17e5))
gen.manual_seed(int(17e5))

strategy = fleet.DistributedStrategy()
fleet.init(is_collective=True,strategy=strategy)
model = MyLayer(emb_dim=args.emb_dim,nlayer=args.nlayer,de_dim=args.de_dim,de_nlayer=args.de_nlayer,drop_path=args.droppath) ###
model.set_state_dict(paddle.load('./log/pretrain/08-15 16:22_29.pdparams'))
model = fleet.distributed_model(model)

model.train()

# training settings
criterion = fts.SoftTargetCrossEntropy()
scheduler = paddle.optimizer.lr.CosineAnnealingDecay(learning_rate=args.lr,T_max=161500, eta_min=8e-6, last_epoch=-1)
warmup = paddle.optimizer.lr.LinearWarmup(scheduler,12000,start_lr=1e-6,end_lr=args.lr)

param_groups = fts.param_groups_lrd(
    model,
        weight_decay=0.05,
        no_weight_decay_list={'pos_embed', 'cls_token', 'dist_token','Wa'},
        layer_decay=.875,
        num_layers=None,
)

optim = paddle.optimizer.AdamW(
    learning_rate = warmup,
    parameters = param_groups,
    weight_decay=args.l2da,
    beta1=0.9, beta2=0.999,
    epsilon = 1e-5)
optim = fleet.distributed_optimizer(optim)

scaler = paddle.amp.GradScaler(init_loss_scaling=1024)

mixup_fn = fts.MixupCutmixHybrid(
    mixup_alpha=args.mixup,
    cutmix_alpha=args.cutmix,
    cutmix_minmax=args.cutmix_minmax,
    prob=args.mixup_prob,
    switch_prob=args.mixup_switch_prob,
    mode=args.mixup_mode,
    label_smoothing=args.smoothing, num_classes=5,
)
# testing settings
criterion_t = paddle.nn.CrossEntropyLoss(
    reduction='mean',
    soft_label=True,
    use_softmax=True,
)
m_t = paddle.metric.Accuracy()

train_ds = HwDataset(mode='train',tail = ds_tail) 
train_sampler = DistributedBatchSampler(train_ds,batch_size=batch_size,shuffle=True)
train_loader = DataLoader(dataset=train_ds,batch_sampler=train_sampler) 

rpt_int = 900000//batch_size if 900000 % batch_size == 0 else 1+ 900000//batch_size
batch_tail = int((len(train_ds)/2/batch_size) % rpt_int)

test_ds = HwDataset(mode='eval')
test_sampler = DistributedBatchSampler(test_ds, batch_size=batch_size,drop_last=True)
test_loader = DataLoader(dataset=test_ds,batch_sampler=test_sampler) 
print('ds prepared in %.f seconds'%(time.time()-ax))

for epoch in range(20):
    for batch_id, x in enumerate(train_loader()):
        model.train()
        attn_input,target,lam = mixup_fn(x[0],x[2])
        with paddle.amp.auto_cast(level='O1'):
            out_k = model._layers.forward_ds(emb_input=x[1],attn_input=attn_input,lam = lam)
            loss_y = criterion(out_k,target) / accu_batch
        loss_p += float(loss_y)
        scaled = scaler.scale(loss_y) #
        scaled.backward()
        if (acc_grad + 1) % accu_batch == 0:
            scaler.minimize(optim,scaled)
            model.clear_gradients()
            warmup.step()
        if (acc_grad + 1) % rpt_int == 0:
            loss_pp = loss_p * accu_batch/rpt_int
            loss_p = 0
            with paddle.no_grad():
                model.eval()
                test_time = 0
                loss_pt = 0
                for test_id,x_t in enumerate(test_loader()):
                    with paddle.amp.auto_cast(level='O1'):
                        out_kt = model._layers.forward_ds(emb_input=x_t[1],attn_input=x_t[0],lam = 1.0)
                        loss_t = criterion_t(out_kt,x_t[2])*x_t[1].shape[0]
                    correct_t = m_t.compute(out_kt,x_t[2])
                    m_t.update(correct_t)
                    res_t = m_t.accumulate()
                    loss_pt += float(loss_t)
                    test_shape += x_t[1].shape[0]
                print('epoch:%.f,'%epoch,time.ctime()[4:19],',loss:%.4f,'%loss_pp,'t_acc:%.3f,'%res_t,'t_loss:%.4f'%(loss_pt/test_shape))

### 其他补充信息 Additional Supplementary Information

log显示时间不一致，并且amp警告时间、内容不一致，怀疑未同步
log_0:
[2023-08-29 09:20:39,386] [    INFO] distributed_strategy.py:160 - distributed strategy initialized
======================= Modified FLAGS detected =======================
FLAGS(name='FLAGS_selected_gpus', current_value='0', default_value='')
=======================================================================
I0829 09:20:39.386621  3239 tcp_utils.cc:181] The server starts to listen on IP_ANY:42982
I0829 09:20:39.386746  3239 tcp_utils.cc:130] Successfully connected to 127.0.1.1:42982
W0829 09:20:39.565924  3239 gpu_resources.cc:119] Please NOTE: device: 0, GPU Compute Capability: 8.6, Driver API Version: 12.2, Runtime API Version: 12.0
W0829 09:20:39.568161  3239 gpu_resources.cc:149] device: 0, cuDNN Version: 8.9.
[2023-08-29 09:20:40,530] [    INFO] topology.py:266 - Total 1 data comm group(s) create successfully!
[2023-08-29 09:20:40,530] [    INFO] topology.py:266 - Total 2 model comm group(s) create successfully!
[2023-08-29 09:20:40,531] [    INFO] topology.py:266 - Total 2 pipe comm group(s) create successfully!
[2023-08-29 09:20:40,531] [    INFO] topology.py:266 - Total 2 sharding comm group(s) create successfully!
[2023-08-29 09:20:40,531] [    INFO] topology.py:217 - HybridParallelInfo: rank_id: 0, mp_degree: 1, sharding_degree: 1, pp_degree: 1, dp_degree: 2, mp_group: [0],  sharding_group: [0], pp_group: [0], dp_group: [0, 1], check/clip group: [0]
/home/johnyan/anaconda3/lib/python3.11/site-packages/paddle/nn/layer/layers.py:1897: UserWarning: Skip loading for pooler.fc_norm.weight. pooler.fc_norm.weight is not found in the provided dict.
  warnings.warn(f"Skip loading for {key}. " + str(err))
/home/johnyan/anaconda3/lib/python3.11/site-packages/paddle/nn/layer/layers.py:1897: UserWarning: Skip loading for pooler.fc_norm.bias. pooler.fc_norm.bias is not found in the provided dict.
  warnings.warn(f"Skip loading for {key}. " + str(err))
ds prepared in 231 seconds
epoch:0, Aug 29 10:08:49 ,loss:1.5297, t_acc:0.312, t_loss:1.4780
epoch:0, Aug 29 10:53:05 ,loss:1.5076, t_acc:0.328, t_loss:1.4719
epoch:0, Aug 29 11:37:22 ,loss:1.5015, t_acc:0.322, t_loss:1.4859
epoch:0, Aug 29 12:21:40 ,loss:1.4983, t_acc:0.329, t_loss:1.4807
epoch:1, Aug 29 13:06:00 ,loss:1.4981, t_acc:0.325, t_loss:1.4732
epoch:1, Aug 29 13:50:16 ,loss:1.4987, t_acc:0.332, t_loss:1.4781
epoch:1, Aug 29 14:34:33 ,loss:1.4981, t_acc:0.333, t_loss:1.4877
epoch:1, Aug 29 15:18:50 ,loss:1.4978, t_acc:0.316, t_loss:1.5069
epoch:1, Aug 29 16:03:07 ,loss:1.4959, t_acc:0.326, t_loss:1.4875
epoch:2, Aug 29 17:07:24 ,loss:1.4938, t_acc:0.329, t_loss:1.5119
epoch:2, Aug 29 17:51:41 ,loss:1.4909, t_acc:0.325, t_loss:1.5252
Found inf or nan, current scale is: 1048576.0, decrease to: 1048576.0*0.5
epoch:2, Aug 29 18:35:57 ,loss:1.4880, t_acc:0.314, t_loss:1.4898
epoch:2, Aug 29 19:20:14 ,loss:1.4866, t_acc:0.301, t_loss:1.5174
Found inf or nan, current scale is: 1048576.0, decrease to: 1048576.0*0.5
epoch:2, Aug 29 20:04:32 ,loss:1.4831, t_acc:0.303, t_loss:1.4897
Found inf or nan, current scale is: 1048576.0, decrease to: 1048576.0*0.5
epoch:3, Aug 29 20:48:51 ,loss:1.4815, t_acc:0.336, t_loss:1.4734
Found inf or nan, current scale is: 524288.0, decrease to: 524288.0*0.5
epoch:3, Aug 29 21:33:08 ,loss:1.4792, t_acc:0.331, t_loss:1.4992
Found inf or nan, current scale is: 524288.0, decrease to: 524288.0*0.5
epoch:3, Aug 29 22:17:24 ,loss:1.4758, t_acc:0.340, t_loss:1.4797
Found inf or nan, current scale is: 524288.0, decrease to: 524288.0*0.5
epoch:3, Aug 29 23:01:41 ,loss:1.4738, t_acc:0.340, t_loss:1.4828
epoch:4, Aug 30 00:05:58 ,loss:1.4700, t_acc:0.334, t_loss:1.5201

log_1:
[2023-08-29 09:20:39,386] [    INFO] distributed_strategy.py:160 - distributed strategy initialized
======================= Modified FLAGS detected =======================
FLAGS(name='FLAGS_selected_gpus', current_value='1', default_value='')
=======================================================================
I0829 09:20:39.386669  3243 tcp_utils.cc:130] Successfully connected to 127.0.1.1:42982
W0829 09:20:39.608963  3243 gpu_resources.cc:119] Please NOTE: device: 1, GPU Compute Capability: 8.6, Driver API Version: 12.2, Runtime API Version: 12.0
W0829 09:20:39.609833  3243 gpu_resources.cc:149] device: 1, cuDNN Version: 8.9.
[2023-08-29 09:20:40,530] [    INFO] topology.py:266 - Total 1 data comm group(s) create successfully!
[2023-08-29 09:20:40,530] [    INFO] topology.py:266 - Total 2 model comm group(s) create successfully!
[2023-08-29 09:20:40,531] [    INFO] topology.py:266 - Total 2 pipe comm group(s) create successfully!
[2023-08-29 09:20:40,531] [    INFO] topology.py:266 - Total 2 sharding comm group(s) create successfully!
[2023-08-29 09:20:40,531] [    INFO] topology.py:217 - HybridParallelInfo: rank_id: 1, mp_degree: 1, sharding_degree: 1, pp_degree: 1, dp_degree: 2, mp_group: [1],  sharding_group: [1], pp_group: [1], dp_group: [0, 1], check/clip group: [1]
/home/johnyan/anaconda3/lib/python3.11/site-packages/paddle/nn/layer/layers.py:1897: UserWarning: Skip loading for pooler.fc_norm.weight. pooler.fc_norm.weight is not found in the provided dict.
  warnings.warn(f"Skip loading for {key}. " + str(err))
/home/johnyan/anaconda3/lib/python3.11/site-packages/paddle/nn/layer/layers.py:1897: UserWarning: Skip loading for pooler.fc_norm.bias. pooler.fc_norm.bias is not found in the provided dict.
  warnings.warn(f"Skip loading for {key}. " + str(err))
ds prepared in 231 seconds
epoch:0, Aug 29 10:08:45 ,loss:1.5287, t_acc:0.320, t_loss:1.4708
epoch:0, Aug 29 10:52:56 ,loss:1.5071, t_acc:0.325, t_loss:1.4700
epoch:0, Aug 29 11:37:08 ,loss:1.5002, t_acc:0.341, t_loss:1.4646
epoch:0, Aug 29 12:21:19 ,loss:1.4971, t_acc:0.317, t_loss:1.4981
epoch:1, Aug 29 13:05:34 ,loss:1.4972, t_acc:0.339, t_loss:1.4692
epoch:1, Aug 29 13:49:48 ,loss:1.4995, t_acc:0.324, t_loss:1.4813
epoch:1, Aug 29 14:34:00 ,loss:1.4988, t_acc:0.317, t_loss:1.4827
epoch:1, Aug 29 15:18:12 ,loss:1.4979, t_acc:0.341, t_loss:1.4729
epoch:1, Aug 29 16:02:25 ,loss:1.4969, t_acc:0.328, t_loss:1.4905
epoch:2, Aug 29 17:06:38 ,loss:1.4932, t_acc:0.330, t_loss:1.4766
epoch:2, Aug 29 17:50:52 ,loss:1.4916, t_acc:0.324, t_loss:1.5003
Found inf or nan, current scale is: 1048576.0, decrease to: 1048576.0*0.5
epoch:2, Aug 29 18:35:06 ,loss:1.4887, t_acc:0.332, t_loss:1.4737
epoch:2, Aug 29 19:19:20 ,loss:1.4868, t_acc:0.337, t_loss:1.4817
epoch:2, Aug 29 20:03:34 ,loss:1.4837, t_acc:0.338, t_loss:1.4745
Found inf or nan, current scale is: 2097152.0, decrease to: 2097152.0*0.5
epoch:3, Aug 29 20:47:48 ,loss:1.4800, t_acc:0.361, t_loss:1.4632
Found inf or nan, current scale is: 2097152.0, decrease to: 2097152.0*0.5
Found inf or nan, current scale is: 1048576.0, decrease to: 1048576.0*0.5
epoch:3, Aug 29 21:32:02 ,loss:1.4781, t_acc:0.348, t_loss:1.4696
Found inf or nan, current scale is: 1048576.0, decrease to: 1048576.0*0.5
epoch:3, Aug 29 22:16:15 ,loss:1.4752, t_acc:0.346, t_loss:1.4810
epoch:3, Aug 29 23:00:27 ,loss:1.4732, t_acc:0.350, t_loss:1.4760
Found inf or nan, current scale is: 1048576.0, decrease to: 1048576.0*0.5
epoch:4, Aug 30 00:04:42 ,loss:1.4704, t_acc:0.337, t_loss:1.4739
[workerlog0.log](https://github.com/PaddlePaddle/Paddle/files/12473905/workerlog0.log)
[workerlog1.log](https://github.com/PaddlePaddle/Paddle/files/12473906/workerlog1.log)
[ftsetting.txt](https://github.com/PaddlePaddle/Paddle/files/12473917/ftsetting.txt)
[hwrun_ft.txt](https://github.com/PaddlePaddle/Paddle/files/12473919/hwrun_ft.txt)
模型使用了droppath：https://github.com/PaddlePaddle/PaddleClas/blob/release/2.5/ppcls/arch/backbone/model_zoo/vision_transformer.py#L71
可能是造成不同步的原因吗？

**Comment by johnyanccer** [comment]
经检查迭代相同步数后的模型参数也不相同（用相同预训练参数初始化模型）
去掉AMP设置仍然不同步模型参数
去掉AMP去掉drop path设置仍然不同步模型参数
有人来帮忙看看吗？

**Comment by johnyanccer** [comment]
找到问题所在了！
model中自定义了forward loop，并且起了别的名字，因此在主程序以及amp decorated部分是形如：
out_kt = model._layers.forward_ds(...)，这样跑就是没能同步模型参数
重新定义model后，使用原关键字forward作为前向程序名称，amp 部分写out_kt = model(...)，于是正常优化，参数同步
